tests/src/Interop/PInvoke/IEnumerator/IEnumeratorNative.h

```diff
@@ -0,0 +1,222 @@
+
+        if (newRefCount == 0)
+        {
+            CoreClrFree(this);
+        }
+
+        return newRefCount;
+    }
+
+    HRESULT STDMETHODCALLTYPE QueryInterface(REFIID iid, void** ppIntf)
+    {
+        HRESULT hr = E_NOINTERFACE;
+        if (iid == __uuidof(IUnknown))
+        {
+            *ppIntf = dynamic_cast<IUnknown*>(this);
+            hr = S_OK;
+        }
+        else if (iid == __uuidof(IEnumVARIANT))
+        {
+            *ppIntf = dynamic_cast<IEnumVARIANT*>(this);
+            hr = S_OK;
+        }
+
+        if (SUCCEEDED(hr))
+        {
+            AddRef();
+        }
+        
+        else
+        {
+            *ppIntf = nullptr;
+        }
+        
+        return hr;
+    }
+};
+
+class IntegerEnumerable : public IDispatch
+{
+private:
+    ULONG refCount = 1;
+    int start;
+    int count;
+public:
+    IntegerEnumerable(int start, int count)
+        :start(start),
+        count(count)
+    {
+    }
+
+    HRESULT STDMETHODCALLTYPE GetTypeInfoCount( 
+        UINT *pctinfo) override
+    {
+        *pctinfo = 0;
+        return S_OK;
+    }
+    
+    HRESULT STDMETHODCALLTYPE GetTypeInfo( 
+        UINT iTInfo,
+        LCID lcid,
+        ITypeInfo **ppTInfo) override
+    {
+        return E_NOTIMPL;
+    }
+    
+    HRESULT STDMETHODCALLTYPE GetIDsOfNames( 
+        REFIID riid,
+        LPOLESTR *rgszNames,
+        UINT cNames,
+        LCID lcid,
+        DISPID *rgDispId) override
+    {
+        bool containsUnknown = false;
+        DISPID *curr = rgDispId;
+        for (UINT i = 0; i < cNames; ++i)
+        {
+            *curr = DISPID_UNKNOWN;
+            LPOLESTR name = rgszNames[i];
+            if(wcscmp(name, W("GetEnumerator")) == 0)
+            {
+                *curr = DISPID_NEWENUM;
+            }
+
+            containsUnknown &= (*curr == DISPID_UNKNOWN);
+            curr++;
+        }
+
+        return (containsUnknown) ? DISP_E_UNKNOWNNAME : S_OK;
+    }
+    
+    HRESULT STDMETHODCALLTYPE Invoke(
+        DISPID dispIdMember,
+        REFIID riid,
+        LCID lcid,
+        WORD wFlags,
+        DISPPARAMS *pDispParams,
+        VARIANT *pVarResult,
+        EXCEPINFO *pExcepInfo,
+        UINT *puArgErr) override
+    {
+        if (dispIdMember == DISPID_NEWENUM && (wFlags & INVOKE_PROPERTYGET) == INVOKE_PROPERTYGET)
+        {
+            V_VT(pVarResult) = VT_UNKNOWN;
+            V_UNKNOWN(pVarResult) = new (CoreClrAlloc(sizeof(IntegerEnumerator))) IntegerEnumerator(start, count);
+            return S_OK;
+        }
+
+        return E_NOTIMPL;
+    }
+    
+    ULONG STDMETHODCALLTYPE AddRef() override
+    {
+        return ++refCount;
+    }
+
+    ULONG STDMETHODCALLTYPE Release() override
+    {
+        ULONG newRefCount = --refCount;
+
+        if (newRefCount == 0)
+        {
+            CoreClrFree(this);
+        }
+
+        return newRefCount;
+    }
+
+    HRESULT STDMETHODCALLTYPE QueryInterface(REFIID iid, void** ppIntf)
+    {
```

Comment:
See the COM tests for implementations of `QueryInterface()`. This is more a question of style, but there is a canonical way to implement `QueryInterface()` as to avoid common bugs.